File: backend_typescript/src/Services/AppKeyRepository.ts
import crypto from 'crypto';
import { sql } from 'kysely';
import { db } from '../db/pools';

/**
 * Mirrors src/AgentTeam/Services/AppKeyRepository.php.
 *
 * CRUD + verification for `app_keys` — long-lived, scoped credentials that
 * let client code call the backend without a user's login JWT.
 *
 * Storage model (see app_keys table):
 *   - key_prefix : first 12 chars of the key ("ak_xxxxxxxx"). Indexed, not
 *                  secret. Used for an O(log N) lookup.
 *   - key_hash   : hex HMAC-SHA256(server_secret, full_key). A DB leak does
 *                  not expose usable keys.
 *
 * The full key value is returned exactly once — by create() — and is never
 * recoverable afterwards.
 */
export class AppKeyRepository {
  private static readonly KEY_BYTES = 16; // 16 random bytes -> 32 hex chars
  private static readonly PREFIX_LEN = 12; // "ak_" + 9 hex chars

  private readonly serverSecret: string;

  constructor(serverSecret: string) {
    if (serverSecret === '') {
      // Fail loud: an empty pepper would make every hash trivially
      // forgeable by anyone who reads the table.
      throw new Error('AppKeyRepository: app_key_secret is not configured.');
    }
    this.serverSecret = serverSecret;
  }

  /**
   * Mint a new app key. Returns row data PLUS `full_key` — the only time the
   * raw key is visible.
   */
  async create(
    userId: number,
    applicationId: string,
    name: string,
    scopes: string[]
  ): Promise<Record<string, any>> {
    const fullKey = 'ak_' + crypto.randomBytes(AppKeyRepository.KEY_BYTES).toString('hex');
    const prefix = fullKey.slice(0, AppKeyRepository.PREFIX_LEN);
    const hash = this.hash(fullKey);

    const res = await sql`
      INSERT INTO app_keys (user_id, application_id, name, key_prefix, key_hash, scopes)
      VALUES (${userId}, ${applicationId}, ${name}, ${prefix}, ${hash}, ${JSON.stringify(scopes)})
    `.execute(db);
    const id = Number((res as any).insertId ?? 0);

    return {
      id,
      user_id: userId,
      application_id: applicationId,
      name,
      key_prefix: prefix,
      scopes: scopes,
      full_key: fullKey,
      created_at: formatSqlDateTime(new Date()),
    };
  }

  /**
   * Verify a presented key. Returns the key row (no hash) or null.
   * Flow: prefix lookup → constant-time hash compare → not-revoked check.
   */
  async findByKey(fullKey: string): Promise<Record<string, any> | null> {
    fullKey = fullKey.trim();
    if (!fullKey.startsWith('ak_') || fullKey.length < AppKeyRepository.PREFIX_LEN) {
      return null;
    }
    const prefix = fullKey.slice(0, AppKeyRepository.PREFIX_LEN);

    const row = (
      await sql<any>`
        SELECT id, user_id, application_id, name, key_prefix, key_hash, scopes,
               created_at, last_used_at, revoked_at
        FROM app_keys
        WHERE key_prefix = ${prefix} AND revoked_at IS NULL
        LIMIT 1
      `.execute(db)
    ).rows[0];
    if (!row) {
      return null;
    }

    if (!timingSafeEqualHex(String(row.key_hash), this.hash(fullKey))) {
      return null;
    }

    delete row.key_hash;
    row.scopes = this.decodeScopes(row.scopes);
    row.id = Number(row.id);
    row.user_id = Number(row.user_id);
    return row;
  }

  /**
   * Stamp last_used_at. Fire-and-forget — a failure here must never block
   * the request that the key just authorized.
   */
  async recordUse(keyId: number): Promise<void> {
    try {
      await sql`UPDATE app_keys SET last_used_at = NOW() WHERE id = ${keyId}`.execute(db);
    } catch (e: any) {
      console.error('[AppKeyRepository] recordUse failed: ' + (e?.message ?? ''));
    }
  }

  /**
   * Soft-delete a key (sets revoked_at). Returns true if a row was changed.
   * Admin-only at the controller layer, so no ownership filter here.
   */
  async revoke(keyId: number): Promise<boolean> {
    const res = await sql`
      UPDATE app_keys SET revoked_at = NOW() WHERE id = ${keyId} AND revoked_at IS NULL
    `.execute(db);
    return Number(res.numAffectedRows ?? 0) > 0;
  }

  /**
   * List keys (admin view). Optional filters. Never returns key_hash.
   */
  async listAll(
    applicationId: string | null = null,
    userId: number | null = null
  ): Promise<Record<string, any>[]> {
    const conds = [sql`1=1`];
    if (applicationId !== null) {
      conds.push(sql`application_id = ${applicationId}`);
    }
    if (userId !== null) {
      conds.push(sql`user_id = ${userId}`);
    }

    const rows = (
      await sql<any>`
        SELECT id, user_id, application_id, name, key_prefix, scopes,
               created_at, last_used_at, revoked_at
        FROM app_keys WHERE ${sql.join(conds, sql` AND `)}
        ORDER BY created_at DESC
      `.execute(db)
    ).rows;

    for (const row of rows) {
      row.id = Number(row.id);
      row.user_id = Number(row.user_id);
      row.scopes = this.decodeScopes(row.scopes);
    }
    return rows;
  }

  /**
   * Fetch one key's metadata by id (no hash). Used by revoke's 404 check.
   */
  async findById(keyId: number): Promise<Record<string, any> | null> {
    const row = (
      await sql<any>`
        SELECT id, user_id, application_id, name, key_prefix, scopes,
               created_at, last_used_at, revoked_at
        FROM app_keys WHERE id = ${keyId} LIMIT 1
      `.execute(db)
    ).rows[0];
    if (!row) {
      return null;
    }
    row.id = Number(row.id);
    row.user_id = Number(row.user_id);
    row.scopes = this.decodeScopes(row.scopes);
    return row;
  }

  private hash(fullKey: string): string {
    return crypto.createHmac('sha256', this.serverSecret).update(fullKey).digest('hex');
  }

  // Mirrors PHP decodeScopes: array passes through; a JSON string is decoded;
  // anything non-array-decoding → []. mysql2 may auto-parse the json column into
  // an array already (see project_mysql2_json_autoparse), so handle both.
  private decodeScopes(raw: any): any[] {
    if (Array.isArray(raw)) {
      return raw;
    }
    try {
      const decoded = JSON.parse(String(raw));
      return Array.isArray(decoded) ? decoded : [];
    } catch {
      return [];
    }
  }
}

/** Constant-time compare of two hex strings (mirrors PHP hash_equals). */
function timingSafeEqualHex(a: string, b: string): boolean {
  const ab = Buffer.from(a);
  const bb = Buffer.from(b);
  if (ab.length !== bb.length) return false;
  return crypto.timingSafeEqual(ab, bb);
}

/** Format a Date as 'Y-m-d H:i:s' (mirrors PHP date('Y-m-d H:i:s')). */
function formatSqlDateTime(d: Date): string {
  const p = (n: number) => String(n).padStart(2, '0');
  return (
    `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ` +
    `${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}`
  );
}

File: backend_python/app/agent_team/services/app_key_repository.py
"""Port of backend/src/AgentTeam/Services/AppKeyRepository.php."""
from __future__ import annotations

import hashlib
import hmac
import json
import secrets

from app.support.logger import error_log
from app.support.phpcompat import php_now

KEY_BYTES = 16
PREFIX_LEN = 12


class AppKeyRepository:
    def __init__(self, db, server_secret: str):
        if server_secret == '':
            raise RuntimeError('AppKeyRepository: app_key_secret is not configured.')
        self.db = db
        self.server_secret = server_secret

    def create(self, user_id: int, application_id: str, name: str, scopes: list) -> dict:
        full_key = 'ak_' + secrets.token_bytes(KEY_BYTES).hex()
        prefix = full_key[:PREFIX_LEN]
        key_id = self.db.insert(
            "INSERT INTO app_keys (user_id, application_id, name, key_prefix, key_hash, scopes)"
            " VALUES (:user_id, :application_id, :name, :key_prefix, :key_hash, :scopes)",
            {':user_id': user_id, ':application_id': application_id, ':name': name,
             ':key_prefix': prefix, ':key_hash': self._hash(full_key),
             ':scopes': json.dumps(list(scopes), separators=(',', ':'))})
        return {'id': key_id, 'user_id': user_id, 'application_id': application_id, 'name': name,
                'key_prefix': prefix, 'scopes': list(scopes), 'full_key': full_key, 'created_at': php_now()}

    def findByKey(self, full_key: str) -> dict | None:
        full_key = full_key.strip()
        if not full_key.startswith('ak_') or len(full_key) < PREFIX_LEN:
            return None
        row = self.db.fetch_one(
            "SELECT id, user_id, application_id, name, key_prefix, key_hash, scopes,"
            " created_at, last_used_at, revoked_at FROM app_keys"
            " WHERE key_prefix = :prefix AND revoked_at IS NULL LIMIT 1",
            {':prefix': full_key[:PREFIX_LEN]})
        if not row:
            return None
        if not hmac.compare_digest(str(row['key_hash']), self._hash(full_key)):
            return None
        row = dict(row)
        del row['key_hash']
        row['scopes'] = self._decode_scopes(row['scopes'])
        row['id'] = int(row['id'])
        row['user_id'] = int(row['user_id'])
        return row

    def recordUse(self, key_id: int) -> None:
        try:
            self.db.execute("UPDATE app_keys SET last_used_at = NOW() WHERE id = ?", [key_id])
        except Exception as e:  # noqa: BLE001
            error_log(f'[AppKeyRepository] recordUse failed: {e}')

    def revoke(self, key_id: int) -> bool:
        return self.db.execute("UPDATE app_keys SET revoked_at = NOW() WHERE id = ? AND revoked_at IS NULL",
                               [key_id]) > 0

    def listAll(self, application_id: str | None = None, user_id: int | None = None) -> list:
        sql = ("SELECT id, user_id, application_id, name, key_prefix, scopes, created_at, last_used_at, revoked_at"
               " FROM app_keys WHERE 1=1")
        params: dict = {}
        if application_id is not None:
            sql += " AND application_id = :application_id"; params[':application_id'] = application_id
        if user_id is not None:
            sql += " AND user_id = :user_id"; params[':user_id'] = user_id
        sql += " ORDER BY created_at DESC"
        rows = self.db.fetch_all(sql, params)
        for r in rows:
            r['id'] = int(r['id']); r['user_id'] = int(r['user_id']); r['scopes'] = self._decode_scopes(r['scopes'])
        return rows

    def findById(self, key_id: int) -> dict | None:
        row = self.db.fetch_one(
            "SELECT id, user_id, application_id, name, key_prefix, scopes, created_at, last_used_at, revoked_at"
            " FROM app_keys WHERE id = ? LIMIT 1", [key_id])
        if not row:
            return None
        row['id'] = int(row['id']); row['user_id'] = int(row['user_id']); row['scopes'] = self._decode_scopes(row['scopes'])
        return row

    def _hash(self, full_key: str) -> str:
        return hmac.new(self.server_secret.encode(), full_key.encode(), hashlib.sha256).hexdigest()

    @staticmethod
    def _decode_scopes(raw):
        """PHP: json_decode($raw, true); is_array($decoded) ? $decoded : []. With
        assoc=true, both JSON arrays and JSON objects decode to a PHP array, so a
        decoded dict is returned as-is (not coerced to a list)."""
        if isinstance(raw, (list, dict)):
            return raw
        try:
            decoded = json.loads(str(raw))
        except ValueError:
            return []
        return decoded if isinstance(decoded, (list, dict)) else []
